grassroots-backend/src/phone-canvass/Scheduler/PhoneCanvassScheduler.ts

```diff
@@ -0,0 +1,153 @@
+  } as const satisfies Record<CallStatus, unknown>;
+  // From caller id.
+  #callers = new Map<number, Caller>();
+
+  #running = false;
+  #pendingContacts: PhoneCanvassContactEntity[];
+
+  metricsTracker: PhoneCanvassMetricsTracker;
+
+  getCurrentTime(): number {
+    return Date.now();
+  }
+
+  constructor(contacts: PhoneCanvassContactEntity[]) {
+    this.#pendingContacts = contacts.filter((contact) => {
+      // TODO(MVP): should this include "UNABLE_TO_CONTACT"?
+      return contact.callStatus === "NOT_STARTED";
+    });
+    this.metricsTracker = new PhoneCanvassMetricsTracker();
+    this.#strategy = new NoOvercallingStrategy(this.metricsTracker);
+  }
+
+  async start(): Promise<void> {
+    if (this.#running) {
+      throw new Error("Call to startScheduler when scheduler already running.");
+    }
+    this.#running = true;
+    while (this.#running) {
+      await this.#strategy.waitForNextCall();
+
+      const contact = this.#pendingContacts.shift();
+      if (contact === undefined) {
+        // We've called all contacts. We're done!
+        this.#callsObservable.complete();
+        break;
+      }
+
+      const notStartedCall = new NotStartedCall({
+        scheduler: this,
+        currentTime: this.getCurrentTime(),
+        contact,
+      });
+      this.callsByStatus.NOT_STARTED.set(notStartedCall.id, notStartedCall);
+      this.#callsObservable.next(notStartedCall);
+      this.metricsTracker.onCallsByStatusUpdate(this.callsByStatus);
+    }
+  }
+
+  stop(): void {
+    this.#running = false;
+  }
+
+  async waitForIdleForTest(): Promise<void> {
+    // To be considered idle we either need no contacts remaining.
+    if (this.#pendingContacts.length === 0) {
+      return;
+    }
+    // or all callers assigned to calls.
+    await firstValueFrom(
+      this.metricsTracker.idleCallerCountObservable.pipe(
+        filter((v) => v === 0),
+      ),
+    );
+  }
+
+  getNextIdleCallerId(): number | undefined {
+    const busyCallerIds = new Set(
+      [...this.callsByStatus.IN_PROGRESS.values()].map((x) => x.callerId),
+    );
+
+    // Find the idle caller who has been available for the longest time.
+    const availableCallers = [...this.#callers.values()].filter((caller) => {
+      return !busyCallerIds.has(caller.id);
+    });
+
+    const firstAvailableCaller = availableCallers.pop();
+    if (firstAvailableCaller === undefined) {
+      return undefined;
+    }
+
+    const oldestAvailableCaller = availableCallers.reduce(
+      (oldest: Caller, current: Caller) => {
+        return oldest.availabilityStartTime < current.availabilityStartTime
+          ? oldest
+          : current;
+      },
+      firstAvailableCaller,
+    );
+
+    return oldestAvailableCaller.id;
+  }
+
+  addCaller(id: number): void {
+    this.#callers.set(id, {
+      id,
+      availabilityStartTime: this.getCurrentTime(),
+    });
+    this.metricsTracker.onCallerCountUpdate(this.#callers.size);
```

Comment:
Rather than have metrics code sprinkled through the call code, would it be possible to have metrics observe the call observables? I know that probably won't always work